backend/zkppSalt.js

```diff
@@ -0,0 +1,229 @@
+    const recordPlaintext = nacl.secretbox.open(recordCiphertext, nonce, encryptionKey)
+
+    if (!recordPlaintext) {
+      return null
+    }
+
+    const recordString = Buffer.from(recordPlaintext).toString('utf-8')
+
+    try {
+      const recordObj = JSON.parse(recordString)
+
+      if (!Array.isArray(recordObj) || recordObj.length !== 3 || !recordObj.reduce((acc, cv) => acc && typeof cv === 'string', true)) {
+        return null
+      }
+
+      const [hashedPassword, authSalt, contractSalt] = recordObj
+
+      return {
+        hashedPassword,
+        authSalt,
+        contractSalt
+      }
+    } catch {
+      // empty
+    }
+  }
+
+  return null
+}
+
+const setZkppSaltRecord = async (contract: string, hashedPassword: string, authSalt: string, contractSalt: string) => {
+  const recordId = blake32Hash(hashStringArray('RID', contract, recordPepper))
+  const encryptionKey = hashStringArray('REK', contract, recordMasterKey).slice(0, nacl.secretbox.keyLength)
+  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength)
+  const recordPlaintext = JSON.stringify([hashedPassword, authSalt, contractSalt])
+  const recordCiphertext = nacl.secretbox(Buffer.from(recordPlaintext), nonce, encryptionKey)
+  const recordBuf = Buffer.concat([nonce, recordCiphertext])
+  const record = recordBuf.toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')
+  await sbp('chelonia/db/set', recordId, record)
+}
+
+export const getChallenge = async (contract: string, b: string): Promise<false | {authSalt: string; s: string; sig: string;}> => {
+  const record = await getZkppSaltRecord(contract)
+  if (!record) {
+    return false
+  }
+  const { authSalt } = record
+  const s = randomNonce()
+  const now = (Date.now() / 1000 | 0).toString(16)
+  const sig = [now, Buffer.from(hashStringArray(contract, b, s, now, challengeSecret)).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')].join(',')
+
+  return {
+    authSalt,
+    s,
+    sig
+  }
+}
+
+const verifyChallenge = (contract: string, r: string, s: string, userSig: string): boolean => {
+  // Check sig has the right format
+  if (!/^[a-fA-F0-9]{1,11},[a-zA-Z0-9_-]{86}(?:==)?$/.test(userSig)) {
+    return false
+  }
+
+  const [then, mac] = userSig.split(',')
+  const now = Date.now() / 1000 | 0
+  const iThen = Number.parseInt(then, 16)
+
+  // Check that sig is no older than Xs
+  if (!(iThen <= now) || !(iThen >= (now - maxAge))) {
+    return false
+  }
+
+  const b = hash(r)
+  const sig = hashStringArray(contract, b, s, then, challengeSecret)
+  const macBuf = Buffer.from(mac.replace(/_/g, '/').replace(/-/g, '+'), 'base64')
+
+  return sig.byteLength === macBuf.byteLength && timingSafeEqual(sig, macBuf)
+}
+
+export const registrationKey = async (contract: string, b: string): Promise<false | {s: string; p: string; sig: string;}> => {
+  const record = await getZkppSaltRecord(contract)
+  if (record) {
+    return false
```

Comment:
Logging, if desired, can be done where the routes are being served. There's deliberately no logging in this protocol primitive implementation.

However, I'd say `console.error` is most likely too high of a logging level. This is a _client_ error, not a _server_ error (i.e., someone trying to do something they shouldn't be doing). As mentioned in other similar comments, zealously logging this can lead to DoS.